### PR TITLE
Add more features to better test move-compiler-v2 support, and remove previous hidden uses of V1 compiler in V2 transactional tests to build precompiled libs.

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Added/changed environment flags to help debugging situations which are useful for isolating and testing compiler v2 support:
- Prevent accidental use of V1:  Adds an env var `MOVE_COMPILER_BLOCK_V1` (`=1` or `=true`) which causes a panic if V1 compiler is used beyond expansion phase.
- Show backtraces for compiler errors for debugging, without necessarily having verbose compiler output: `MOVE_COMPILER_BACKTRACE` (`=1` or `=true`).  Note that you also need to set `RUST_BACKTRACE` to actually get anything.
- Aliases for these and various other `MOVE_COMPILER_XYZ` flags as `MVC_XYZ` to please slow typists.  (`MVC_V2`, `MVC_BLOCK_V1`, `MVC_BACKTRACE`, `MVC_DEBUG`).
- Loosens requirement on `move-prover` to allow a value of `1` (in addition to `"true"`) to enable `MOVE_COMPILER_V2`.

Fix transactional tests (for third_party and also aptos) to be able to avoid V1 compiler:
- Remove uses of (v1 compiled) precompiled libs in V2 transactional tests.
- If `MOVE_COMPILER_V2=true`, avoid comparison testing and exclusively use V2 compiler.

In summary: to see runs without V1 use try
```
    MOVE_COMPILER_V2=true MOVE_COMPILER_BLOCK_V1=true UB=true cargo test -p move-compiler-v2-transactional-tests -p aptos-transactional-test-harness ...
```

## Usage

Add
```
    MOVE_COMPILER_BACKTRACE=1 RUST_BACKTRACE=1
```
if you want to see backtraces from panics and compiler bugs.

## Tested
Extensive testing with the various environment variables implemented

With no special variables defined, test output should be unchanged from before.

Currently, certain tests fail with MVC_V2=true MVC_BLOCK_V1=true due to extended_checks and/or test_validation::needs_validation().  Hopefully Teng's fix to extended_checks will solve these:
```
failures:
    move_aptos_stdlib_unit_tests
    move_framework_unit_tests
    move_stdlib_unit_tests
    move_token_objects_unit_tests
    move_token_unit_tests
```

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [x] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)
   move-compiler-v2

## Key Areas to Review
Try it out, maybe.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-prover/move-docgen/tests/testsuite.rs [hunk 1]
```diff
@@ -76,6 +76,7 @@ fn test_docgen(path: &Path, mut options: Options, suffix: &str) -> anyhow::Resul
     let mut temp_path = PathBuf::from(TempDir::new()?.path());
     options.docgen.output_directory = temp_path.to_string_lossy().to_string();
     options.skip_attribute_checks = true;
+    options.prover.dump_bytecode = true;
```
Comment (rahxephon89) [inline third_party/move/move-prover/move-docgen/tests/testsuite.rs:79]: What is this option used for?

File: third_party/move/move-prover/bytecode-pipeline/src/number_operation.rs [hunk 2]
```diff
@@ -217,6 +217,14 @@ impl GlobalNumberOperationState {
             } else {
                 // If not appearing in the pragma, mark it as Arithmetic or Bottom
                 // Similar logic when populating ret_operation_map below
+                let opt_local_type = func_env.get_local_type(i);
```
Comment (rahxephon89) [inline third_party/move/move-prover/bytecode-pipeline/src/number_operation.rs:220]: This change is orthogonal to other changes in this PR, right?

File: third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 3]
```diff
@@ -463,19 +481,59 @@ pub fn run_test(path: &Path) -> Result<(), Box<dyn std::error::Error>> {
...
+    }
+}
+
 pub fn run_test_with_config(
-    config: TestRunConfig,
+    mut config: TestRunConfig,
     path: &Path,
 ) -> Result<(), Box<dyn std::error::Error>> {
-    run_test_impl::<SimpleVMTestAdapter>(config, path, Some(&*PRECOMPILED_MOVE_STDLIB), &None)
+    if read_bool_env_var(MOVE_COMPILER_V2) && !matches!(config, TestRunConfig::CompilerV2 { .. }) {
```
Comment (rahxephon89) [inline third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs:508]: Do we have to use the environment variable here?

File: third_party/move/move-prover/bytecode-pipeline/src/number_operation.rs (hunk 2 above)
Comment (brmataptos) [reply third_party/move/move-prover/bytecode-pipeline/src/number_operation.rs:220]: Oops, I should maybe remove this one.

File: third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs (hunk 3 above)
Comment (brmataptos) [reply third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs:508]: This is the entry point from transactional-tests under various dirs (move-bytecode-verifier, move-ir-compiler, move-vm), so if we want to use V2 in those tests we need to have the check somewhere, and we need it before the precompiled V1 libs are compiled.

File: third_party/move/move-compiler/src/naming/translate.rs [hunk 4]
```diff
@@ -353,6 +353,9 @@ pub fn program(
     prog: E::Program,
 ) -> N::Program {
     let mut context = Context::new(compilation_env, pre_compiled_lib, &prog);
+    if context.env.flags().get_block_v1_compiler() {
```
Comment (vineethk) [inline third_party/move/move-compiler/src/naming/translate.rs:356]: I really like this idea - simple, but would have saved me some time wondering "are we sure the compiler v2 ran" in a few circumstances.

File: third_party/move/move-prover/src/lib.rs [hunk 5]
```diff
@@ -138,6 +138,7 @@ pub fn run_move_prover_with_model_v2<W: WriteColor>(
     options: Options,
     start_time: Instant,
 ) -> anyhow::Result<()> {
+    eprintln!("global env before prover run:\n{}", env.dump_env_all());
```
Comment (vineethk) [inline third_party/move/move-prover/src/lib.rs:141]: remove.

File: third_party/move/move-prover/src/lib.rs [hunk 6]
```diff
@@ -282,6 +283,7 @@ pub fn create_and_process_bytecode(options: &Options, env: &GlobalEnv) -> Functi
         Path::new(s).file_name().unwrap().to_str().unwrap()
     });
 
+    eprintln!("create_and_process_bytecode");
```
Comment (vineethk) [inline third_party/move/move-prover/src/lib.rs:286]: remove.

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs [hunk 7]
```diff
@@ -613,7 +636,7 @@ impl<'a> CompiledState<'a> {
         module: CompiledModule,
         source_file: (String, NamedTempFile),
     ) {
-        let id = module.self_id();
+        let id: ModuleId = module.self_id();
```
Comment (vineethk) [inline third_party/move/testing-infra/transactional-test-runner/src/framework.rs:639]: remove type annotation?

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs [hunk 8]
```diff
@@ -675,29 +698,26 @@ impl<'a> CompiledState<'a> {
...
     experiments: Vec<(String, bool)>,
-) -> Result<(
-    (Option<CompiledModule>, Option<CompiledScript>),
-    Option<String>,
-)> {
+) -> Result<(AnnotatedCompiledUnit, Option<String>)> {
     let deps = if let Some(p) = pre_compiled_deps {
-        // The v2 compiler does not (and perhaps never) supports precompiled programs, so
-        // compile from the sources again, computing the directories where they are found.
+        // The v2 compiler does not really support precompiled programs, so we must include all the
```
Comment (vineethk) [inline third_party/move/testing-infra/transactional-test-runner/src/framework.rs:715]: Not an action item for this PR, but wanted to understand:
(1) why does v2 compiler not really support precompiled programs (whereas, I assume, v1 does)?
(2) can this be a source of performance inefficiencies when using v2?

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs [hunk 9]
```diff
@@ -732,10 +752,11 @@ fn compile_source_unit_v2(
     let unit = if units.len() != 1 {
         anyhow::bail!("expected either one script or one module")
     } else {
-        match units.pop().unwrap() {
-            AnnotatedCompiledUnit::Module(m) => (Some(m.named_module.module), None),
-            AnnotatedCompiledUnit::Script(s) => (None, Some(s.named_script.script)),
-        }
+        units.pop().unwrap()
+        // match units.pop().unwrap() {
```
Comment (vineethk) [inline third_party/move/testing-infra/transactional-test-runner/src/framework.rs:756]: remove commented code

File: third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 10]
```diff
@@ -407,43 +425,43 @@ static PRECOMPILED_MOVE_STDLIB: Lazy<FullyCompiledProgram> = Lazy::new(|| {
...
-                AnnotatedCompiledUnit::Module(annot_module) => {
-                    Some(annot_module.named_module.module)
-                },
-                AnnotatedCompiledUnit::Script(_) => None,
-            })
-            .collect(),
+pub struct PrecompiledFilesModules(Vec<String>, Vec<AnnotatedCompiledUnit>);
+
+impl PrecompiledFilesModules {
+    pub fn new(files: Vec<String>, modules: Vec<AnnotatedCompiledUnit>) -> Self {
```
Comment (vineethk) [inline third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs:439]: I am assuming `files` are vector of strings because of the type plumbing, but seems like they might be better as vector of paths (from just seeing this struct). This is a nit.

File: third_party/move/move-prover/tests/sources/regression/vector_theory_boogie_array.v2_exp [hunk 11]
```diff
@@ -13,7 +13,7 @@ error: post-condition does not hold
    =         addr = <redacted>
    =     at tests/sources/regression/vector_theory_boogie_array.move:7: f1
    =     at tests/sources/regression/vector_theory_boogie_array.move:8: f1
-   =         pool = <redacted>
+   =         idx = <redacted>
```
Comment (vineethk) [inline third_party/move/move-prover/tests/sources/regression/vector_theory_boogie_array.v2_exp:16]: I don't understand this baseline change - I am assuming it has been vetted.

**Review by vineethk — COMMENTED** [review]
Looking good so far. Have a few comments.

File: third_party/move/move-prover/tests/sources/regression/vector_theory_boogie_array.v2_exp (hunk 11 above)
Comment (rahxephon89) [reply third_party/move/move-prover/tests/sources/regression/vector_theory_boogie_array.v2_exp:16]: This is an issue from the prover when generating the trace, which depends on the boogie output as far as I remember.

**Review by brmataptos — COMMENTED** [review]
I've updated the description as well as the PR to address comments.  Baseline (no environment variables set) .exp files are now unchanged.

**Review by vineethk — APPROVED** [review]

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs (hunk 9 above)
Comment (rahxephon89) [reply third_party/move/testing-infra/transactional-test-runner/src/framework.rs:756]: @brmataptos The commented code has not been removed yet.

**Review by rahxephon89 — APPROVED** [review]

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs (hunk 9 above)
Comment (brmataptos) [reply third_party/move/testing-infra/transactional-test-runner/src/framework.rs]: Done now.

**Review by wrwg — APPROVED** [review]
